### ISSUE TITLE
Add option --skip-my-cnf to prevent automatic parsing of ~/.my.cnf

### ISSUE THREAD

**Issue by d33psky** [skeema/skeema#64]
To support use of multiple mysql.cnf files, or to force to use none of the regular .my.cnf settings by pointing to an empty one.

ps. should also prove useful to work around this issue : 
`2019-03-14 13:06:23 [ERROR] Option host cannot be set via .my.cnf for this command`
but that's less important.

**Comment by evanelias** [comment]
> 2019-03-14 13:06:23 [ERROR] Option host cannot be set via .my.cnf for this command

Ah, that's actually a bug. Thanks for reporting this! I'll put a fix in the next patch release, currently aiming for tomorrow. `host` in ~/.my.cnf will be ignored by all commands.

Separate from that though, can you explain more your use-case for wanting --defaults-file? Skeema already has a hierarchy of global option files; see https://github.com/skeema/skeema/blob/master/doc/config.md#specifying-options-via-option-files.

If your use-case is wanting to supply different username/password in various situations, I'd recommend using different environment sections in ~/.skeema, e.g.:

```ini
[production]
user=admin
password=123

[dev]
user=root
password=456
```

**Comment by evanelias** [comment]
A fix for `host` in ~/.my.cnf is now in master. I was hoping to make a new Skeema release today, but my CI vendor seems to be broken yet again, so it may need to wait until Monday.

**Comment by d33psky** [comment]
> > 2019-03-14 13:06:23 [ERROR] Option host cannot be set via .my.cnf for this command
> Ah, that's actually a bug.

That's what I hoped.

> Thanks for reporting this! I'll put a fix in the next patch release, currently aiming for tomorrow. `host` in ~/.my.cnf will be ignored by all commands.

> Separate from that though, can you explain more your use-case for wanting --defaults-file? Skeema already has a hierarchy of global option files; see https://github.com/skeema/skeema/blob/master/doc/config.md#specifying-options-via-option-files.

Yes but I find that insufficient for trust reasons. Instead of a --defaults file option I would be happy with a --no-defaults flag (just like mysql has) to make absolutely sure that skeema is not using *anything* from a .my.cnf it may encounter.

**Comment by evanelias** [comment]
Thanks. If you could explain your use-case in more depth, it will help me evaluate and prioritize this.

Would you want --no-defaults to only skip ~/.my.cnf? Or all home dir option files (~/.my.cnf and ~/.skeema)? Or all global option files altogether (those 2 + /etc/skeema and /usr/local/etc/skeema)?

Are there specific options where you are concerned about Skeema using the value in ~/.my.cnf?

**Comment by d33psky** [comment]
> Would you want --no-defaults to only skip `~/.my.cnf`?

Yes. Ideally skeema behaves a bit like the mysql client does.
I need to be completely sure where skeema gets its config from.
So if skeema can use settings from .my.cnf I want to be able to tell it which one to use.
If that is not possible I want to be able to decouple .my.cnf from skeema completely with something like a --no-defaults option.
I have multiple my.cnf files for various databases and would like to be able to use skeema on a few of them.

> Or all home dir option files (`~/.my.cnf` and `~/.skeema`)? Or all global option files altogether (those 2 + `/etc/skeema` and `/usr/local/etc/skeema`)?

That sounds useful too. But is not what I need at this point.
If you choose to implement this then maybe distinguish with names like
- --no-my-cnf-defaults
- --no-skeema-defaults
?

> Are there specific options where you are concerned about Skeema using the value in ~/.my.cnf?

The hosts entry was one ;-)
But the thing is that it needs to be absolutely clear what config skeema uses.

**Comment by evanelias** [comment]
> I have multiple my.cnf files for various databases and would like to be able to use skeema on a few of them.

Is this re: auth info (`user` and `password` options) specifically, or something else?

If you just want to configure different auth info on different DBs, the way to do this in Skeema is via named environments. And since `~/.skeema` takes precedence over `~/.my.cnf`,  there's a pre-existing way to ensure you're overriding things in `~/.my.cnf`.

I realize that doesn't solve your root concern adequately, but it may provide a sufficient work-around for you for now.

> If you choose to implement this then maybe distinguish with names like

I'd lean more towards just a single option (to minimize complexity), skipping .my.cnf only.

Or maybe an alternative approach of not adding any new option, and instead just restricting which options are ever parsed from `~/.my.cnf`'s [client] and [mysql] sections to just a documented whitelist of `user`, `password`, and perhaps `socket`. 

For backwards-compatibility, we would need to continue parsing all options from the [skeema] section in ~/.my.cnf. That seems more straightforward/expected though.

> The hosts entry was one ;-)

To be clear, even in versions affected by this bug, presence of `host` in `~/.my.cnf` was caught by a safety check and errored loudly. It's not like the wrong value was ever silently used in a detrimental manner.

> But the thing is that it needs to be absolutely clear what config skeema uses.

This makes sense conceptually, though practically speaking I'm having difficulty seeing the harm vector. e.g. if Skeema picks up the wrong user/pass combo from ~/.my.cnf you'll just get an auth error.

**Comment by d33psky** [comment]
> > I have multiple my.cnf files for various databases and would like to be able to use skeema on a few of them.
> Is this re: auth info (`user` and `password` options) specifically, or something else?

Not specifically and nothing particular else either. I just want to enforce that nothing from a self-probed config file is used at all. (with self-probed I mean that I did not explicitly tell skeema to use a .my.cnf)

> If you just want to configure different auth info on different DBs, the way to do this in Skeema is via named environments. And since `~/.skeema` takes precedence over `~/.my.cnf`, there's a pre-existing way to ensure you're overriding things in `~/.my.cnf`.

Is that on per-key-value basis ? If so that does not guarantee that my override settings are complete.

> I realize that doesn't solve your root concern adequately, but it may provide a sufficient work-around for you for now.

Can I make an empty `~/.skeema` to invalidate any possible `~/.my.cnf` ?

> I'd lean more towards just a single option (to minimize complexity), skipping .my.cnf only.

That'd be great :)

> Or maybe an alternative approach of not adding any new option, and instead just restricting which options are ever parsed from `~/.my.cnf`'s [client] and [mysql] sections to just a documented whitelist of `user`, `password`, and perhaps `socket`.

That would also work, but I don't like it.

> This makes sense conceptually, though practically speaking I'm having difficulty seeing the harm vector. e.g. if Skeema picks up the wrong user/pass combo from ~/.my.cnf you'll just get an auth error.

The harm vector is obvious to me: manipulate the wrong database. But mixing up credentials is unwanted as well.

Concluding I think a `--no-my-cnf-defaults` would work best.
I hope this helps.

**Comment by evanelias** [comment]
> Is that on per-key-value basis ?

Yes.

> Can I make an empty ~/.skeema to invalidate any possible ~/.my.cnf ?

No. You'd need to override the specific keys.

> The harm vector is obvious to me: manipulate the wrong database. 

That is not possible. Skeema [explicitly does not permit](https://github.com/skeema/skeema/blob/master/doc/config.md#limitations-on-host-and-schema-options) `host` or `schema` to be configured in a global configuration file. There is no risk of values in my.cnf causing Skeema to manipulate the wrong database.

> But mixing up credentials is unwanted as well.

Again, worst case there is Skeema is unable to connect to the database. 

> Concluding I think a --no-my-cnf-defaults would work best.

OK. I'll keep this open in case a third-party wants to contribute code for it.

**Comment by evanelias** [comment]
Just released [Skeema v1.2.3](https://github.com/skeema/skeema/releases/tag/v1.2.3), which includes new option [skip-my-cnf](https://github.com/skeema/skeema/blob/master/doc/options.md#my-cnf).
